### ISSUE TITLE
Disable Menu Opening on Hover

### ISSUE THREAD

**Issue by SephLeague** [ocornut/imgui#1416]
Hello

Is it possible to disable the automatic opening of the menu if its been hovered? When you hover a menu created with BeginMenu, it automatically closes other menus that may be open and opens this one. 
This is good but i would like it to only do this behaviour if its been clicked and not simply on hovering. 

Thanks

**Comment by ocornut** [comment]
Hello,
It's not possible at the moment. The current behavior is modelled after what OSes such as Windows do.
Could you clarify why you are interested in changing this behavior? I don't know of a sensible reason at the moment to justify adding extra complexity to the existing code.

**Comment by SephLeague** [comment]
Well the main reason is because I was using the Menu as the main gui to traverse through horizontally instead of as a menubar. The reason I was using this is because it expands horizontally where as the treenodes and collapsing headers just go down vertically and things get too compacted. But since my case is pretty unique, i guess its not too important, i just wondered if the functionality is already existing.

**Comment by ocornut** [comment]
So but I understand correctly, you have multiple calls to `BeginMenu()` outside of a menubar?
Like:
![image](https://user-images.githubusercontent.com/8225057/32469181-c9450214-c351-11e7-8ed1-0d2e21ab6187.png)

I think in this situation there is an unhandled case, the bit of code in BeginMenu that handle this algorithm:
  http://bjk5.com/post/44698559168/breaking-down-amazons-mega-dropdown
Doesn't run at all in this specific use case. This is because the code is detecting sibling menus based on parent popups and there's no parent popups. Without that code it is hard to move the mouse toward the sub-menu because you can easily accidentally hover the menu above or below.
You'll notice that within a regular menu you won't have the problem.

TL;DR; _if_ I understad your problem correctly, there is bug that makes the interaction worse that it should be, and it may be possible to fix the bug.

**Comment by SephLeague** [comment]
Yes. I have the calls outside of the menubar. As i said the main purpose is i want a GUI like with the treenodes and collapsing headers except I want it to be horizontal instead of vertical. So i resorted to this but it doesnt match my criteria because with a tree node or collapsing header, you can toggle it on and off to change its visible state. With the menu, i hoped to achieve a similiar affect but horizontally but obviously its meant for Menubars and toggling them on and off is not standard behaviour. 

So yes, in the screenshot you linked, when you click the "Menu inside a regular window" I  want it to expand out and in just like a collapsing header.

**Comment by ocornut** [comment]
OK so ignore what I posted earlier, the triangle code is actually functional in that situation, it's just much easier for it to fail considering the BeginMenu item width tends to be larger.

The logic change to get what you want is easy to write but deciding on which criteria to enable that logic is trickier. Test patch posted below. I think it makes things relatively inconsistent to enable it with that logic because as you get into a sub-menu it starts behaving differently.

I expect to add flags to BeginMenu (#1379 needs it) we can perfectly make that behavior a new flag to `BeginMenu()`.

```
From b85631e9f44a18c27f70c8c5209ad3e94d3b39a6 Mon Sep 17 00:00:00 2001
From: omar <omarcornut@XXX.com>
Date: Tue, 7 Nov 2017 00:54:01 +0100
Subject: [PATCH] Menus: testing specific click behavior for menus inside
 windows (#1416)

---
 imgui.cpp | 55 ++++++++++++++++++++++++++++++++++++-------------------
 1 file changed, 36 insertions(+), 19 deletions(-)

diff --git a/imgui.cpp b/imgui.cpp
index 9c9e7b1..c8ed22a 100644
--- a/imgui.cpp
+++ b/imgui.cpp
@@ -9221,33 +9221,50 @@ bool ImGui::BeginMenu(const char* label, bool enabled)
         g.NavWindow = backed_nav_window;
 
     bool want_open = false, want_close = false;
-    if (window->DC.LayoutType != ImGuiLayoutType_Horizontal) // (window->Flags & (ImGuiWindowFlags_Popup|ImGuiWindowFlags_ChildMenu))
+    if (window->DC.LayoutType == ImGuiLayoutType_Vertical) // (window->Flags & (ImGuiWindowFlags_Popup|ImGuiWindowFlags_ChildMenu))
     {
-        // Implement http://bjk5.com/post/44698559168/breaking-down-amazons-mega-dropdown to avoid using timers, so menus feels more reactive.
-        bool moving_within_opened_triangle = false;
-        if (g.HoveredWindow == window && g.OpenPopupStack.Size > g.CurrentPopupStack.Size && g.OpenPopupStack[g.CurrentPopupStack.Size].ParentWindow == window)
+        if (window->Flags & ImGuiWindowFlags_ChildMenu)
         {
-            if (ImGuiWindow* next_window = g.OpenPopupStack[g.CurrentPopupStack.Size].Window)
+            // Regular path: Menus inside Menus
+            // Implement http://bjk5.com/post/44698559168/breaking-down-amazons-mega-dropdown to avoid using timers, so menus feels more reactive.
+            bool moving_within_opened_triangle = false;
+            if (g.HoveredWindow == window && g.OpenPopupStack.Size > g.CurrentPopupStack.Size && g.OpenPopupStack[g.CurrentPopupStack.Size].ParentWindow == window)
             {
-                ImRect next_window_rect = next_window->Rect();
-                ImVec2 ta = g.IO.MousePos - g.IO.MouseDelta;
-                ImVec2 tb = (window->Pos.x < next_window->Pos.x) ? next_window_rect.GetTL() : next_window_rect.GetTR();
-                ImVec2 tc = (window->Pos.x < next_window->Pos.x) ? next_window_rect.GetBL() : next_window_rect.GetBR();
-                float extra = ImClamp(fabsf(ta.x - tb.x) * 0.30f, 5.0f, 30.0f); // add a bit of extra slack.
-                ta.x += (window->Pos.x < next_window->Pos.x) ? -0.5f : +0.5f;   // to avoid numerical issues
-                tb.y = ta.y + ImMax((tb.y - extra) - ta.y, -100.0f);            // triangle is maximum 200 high to limit the slope and the bias toward large sub-menus // FIXME: Multiply by fb_scale?
-                tc.y = ta.y + ImMin((tc.y + extra) - ta.y, +100.0f);
-                moving_within_opened_triangle = ImTriangleContainsPoint(ta, tb, tc, g.IO.MousePos);
-                //window->DrawList->PushClipRectFullScreen(); window->DrawList->AddTriangleFilled(ta, tb, tc, moving_within_opened_triangle ? IM_COL32(0,128,0,128) : IM_COL32(128,0,0,128)); window->DrawList->PopClipRect(); // Debug
+                if (ImGuiWindow* next_window = g.OpenPopupStack[g.CurrentPopupStack.Size].Window)
+                {
+                    ImRect next_window_rect = next_window->Rect();
+                    ImVec2 ta = g.IO.MousePos - g.IO.MouseDelta;
+                    ImVec2 tb = (window->Pos.x < next_window->Pos.x) ? next_window_rect.GetTL() : next_window_rect.GetTR();
+                    ImVec2 tc = (window->Pos.x < next_window->Pos.x) ? next_window_rect.GetBL() : next_window_rect.GetBR();
+                    float extra = ImClamp(fabsf(ta.x - tb.x) * 0.30f, 5.0f, 30.0f); // add a bit of extra slack.
+                    ta.x += (window->Pos.x < next_window->Pos.x) ? -0.5f : +0.5f;   // to avoid numerical issues
+                    tb.y = ta.y + ImMax((tb.y - extra) - ta.y, -100.0f);            // triangle is maximum 200 high to limit the slope and the bias toward large sub-menus // FIXME: Multiply by fb_scale?
+                    tc.y = ta.y + ImMin((tc.y + extra) - ta.y, +100.0f);
+                    moving_within_opened_triangle = ImTriangleContainsPoint(ta, tb, tc, g.IO.MousePos);
+                    //window->DrawList->PushClipRectFullScreen(); window->DrawList->AddTriangleFilled(ta, tb, tc, moving_within_opened_triangle ? IM_COL32(0,128,0,128) : IM_COL32(128,0,0,128)); window->DrawList->PopClipRect(); // Debug
+                }
             }
-        }
 
-        want_close = (menu_is_open && !hovered && g.HoveredWindow == window && g.HoveredIdPreviousFrame != 0 && g.HoveredIdPreviousFrame != id && !moving_within_opened_triangle);
-        want_open = (!menu_is_open && hovered && !moving_within_opened_triangle) || (!menu_is_open && hovered && pressed);
+            want_close = (menu_is_open && !hovered && g.HoveredWindow == window && g.HoveredIdPreviousFrame != 0 && g.HoveredIdPreviousFrame != id && !moving_within_opened_triangle);
+            want_open = (!menu_is_open && hovered && !moving_within_opened_triangle) || (!menu_is_open && hovered && pressed);
+        }
+        else
+        {
+            // Menus inside Windows
+            if (menu_is_open && pressed && menuset_is_open) // Click an open menu again to close it
+            {
+                want_close = true;
+                want_open = menu_is_open = false;
+            }
+            else if (pressed)
+            {
+                want_open = true;
+            }
+        }
     }
     else
     {
-        // Menu bar
+        // Menus inside Menu Bar
         if (menu_is_open && pressed && menuset_is_open) // Click an open menu again to close it
         {
             want_close = true;
-- 
2.10.2.windows.1

```

**Comment by SephLeague** [comment]
Ok. First of all, thank for all the hard work you put in. The library is pretty amazing and I'm loving it. 

"I think it makes things relatively inconsistent to enable it with that logic because as you get into a sub-menu it starts behaving differently." 

Not entirely sure what you mean by that but if i were to apply that, i would apply it to all the children menus too so that it is consistent. 

Essentially, all i want is horizontal collapsing headers and treenodes. 

Thank you

**Comment by markitus18** [comment]
Hello.
I am having the same problem, i call BeginMenu() outside of a MenuBar() and i do not want it to open on hover.
I've been searching though ImGui versions and an older version had the behaviour i expected.
I guess changing the current version lines for the older ones would work as a temporal solution. I need to test it now


Current Version
    ImGuiWindow* backed_nav_window = g.NavWindow;
    if (menuset_is_open)
        g.NavWindow = window;  // Odd hack to allow hovering across menus of a same menu-set 
                                               (otherwise we wouldn't be able to hover parent)

Older Version:
    ImGuiWindow* backed_focused_window = g.FocusedWindow;
    [....]
    if (menuset_is_open)
        g.FocusedWindow = window;

**Comment by reliasn** [comment]
I just updated from 1.50 to 1.63 and I can confirm this behavior.

On 1.63, simply hovering the menu and sub-menus opens them.
On 1.50, you first need to click the menu to open it. Sub-menus are opened on hover.

@markitus18 's solution makes the menu and sub-menus to be opened only on clicks, so the navigation isn't fast as a simple hover.

@ocornut 's solution does the job, although the diff shown is a bit confusing. In a sum, add this check in order to get 1.50's behavior:
```c
if (window->Flags & ImGuiWindowFlags_ChildMenu) {
// old code
} else {
    if (pressed)
    {
        want_open = true;
    }
}
```
The other ``if`` in ocornut's code isn't needed.

**Comment by SirLennox** [comment]
bump, can't disable opening menu on hover yet, please add. @ocornut
